### ISSUE TITLE
missing textures: bug or missing in original W3D model?

### ISSUE THREAD

**Issue by ghost** [tomysshadow/Shockwave-3D-World-Converter#4]
Hi

I tried to convert these models here to OBJ and imported them into Blender (https://www.blender.org/):

- Download the file https://www.mameworld.info/3darcade/models/sw3d/1942model%20v1.zip

I imported the OBJ into Blender and have this result:

![afbeelding](https://user-images.githubusercontent.com/122493908/211935893-07adc574-050a-42ac-b985-233cb2ceb3bd.png)

It should look like (purple parts are missing textures):

https://www.mameworld.info/3darcade/models/pic/1942.jpg

The MTL file shows TIFF textures which are not saved.

- The second try for another model.

Download: https://www.mameworld.info/3darcade/models/sw3d/simpsonsmodel%20v1.zip

In Blender: 
![afbeelding](https://user-images.githubusercontent.com/122493908/211936214-3c544989-0671-421f-ba74-9405491bc617.png)

It should look like (purple parts are missing textures):

https://www.mameworld.info/3darcade/models/pic/simpsons.jpg

Is there a bug that made the program not save all TIFF files?
Or are those models just not containing the missing textures?
The models were originally used in a program called 3D Arcade so I think there is only a little possibility the textures are missing in the W3D file (but you never know for sure).

Can you please have a look at it? I'm a beginner and don't know anything about W3D. I just use your cool tool!

Thanks!

**Comment by ghost** [comment]
You can try more W3D models here

https://www.mameworld.info/3darcade/

Click on "models" and search for W3D

**Comment by tomysshadow** [comment]
It would appear that this is indeed a bug. The materials in question are intended to be solid colours, and aren't supposed to have images at all. However, the `useDiffuseWithTexture` property of the shader is false, so it tries to write out a map_Kd statement instead of a regular Kd. I guess there needs to be an exception where if there are no images, to use the diffuse colour even if `useDiffuseWithTexture` is false. Will fix.

**Comment by tomysshadow** [comment]
Tested with the 1942 model as well as my regular test models, should be patched in 1.3.1.

**Comment by ghost** [comment]
Thanks a lot for the quick update!

It works now without any problems! I'm testing many models so I let you know if I find another problem.

Can I ask following questions?

- I know OBJ is an open format, but I don't think W3D is. So how do you know how to read it?

- In what programming language is this program written because there is no source code here?

- Are textures in a W3D model always TIFF format or did you just choose this format? Can I for example request also JPG format as a new feature?

**Comment by tomysshadow** [comment]
The source code is in fact here, you just need to open the project file
(Shockwave3DWorldConverter.dir) in Adobe Director 11.5, and you will be
able to view it.

It's written in Lingo (the scripting language that Shockwave uses.) How do
I know how to read the format - well, I just read the Adobe documentation
;) in Shockwave 3D, you can use the `meshDeform` modifier to access or
modify the positions of individual faces and vertices in a mesh. So I just
wrote a program that uses the existing Shockwave 3D Xtra to loop through
all the meshes, read that data, then write it out to a file.

Basically, because Shockwave 3D allows you to access the mesh deform
through its scripting language, I didn't need to reverse the actual format.
Unfortunately, this seems to stop at bones which aren't exposed by the API,
but it might be possible to write another Xtra that can export those
eventually.

**Comment by ghost** [comment]
Thanks!
Now I understand it but Adobe Director 11.5 seems to be abandonware?

**Comment by tomysshadow** [comment]
Adobe Director (formerly Macromedia Director) is the authoring tool used to
create Shockwave Movies. The DIR, DXR, and DCR filetypes are created by it
(the former for project files, the latter two for compiled results.) It was
discontinued in 2016 but you can still find it secondhand and does mostly
work on Windows 10 so long as you have an activation key. There's a bug
that's cropped up in recent years where certain Nvidia drivers cause it to
crash, but this project uses an Xtra which automatically fixes it.

It definitely won't work with an older version of Director but might also
open in Director 12, the final version, if you're able to find it instead
(though it needs to be registered online, similar to Adobe CS6.)

Unfortunately the Shockwave 3D World format is sufficiently advanced that
Wavefront OBJ doesn't support all of its features. For example, texture
layers - currently I work around this limitation by exporting layered
textures as multipage TIFF, and the expectation is you'll use an image
editor to break it into the individual layers and reassign them, but this
is not ideal: they really should be separate image files and specified as
layers in the model itself.

I at some future point wanted to add support for Autodesk FBX which does
support such advanced features. (Shockwave 3D World files are exported
from 3DS Max, so FBX would be the closest analogue.) It'd require writing
an Xtra to wrap the FBX SDK so that I can control it through Lingo, which
is a lot more effort than just writing them directly to a plaintext file. If I do
this at some point, it'd be natural to add support for bones at the same time.

**Comment by ghost** [comment]
I don't know much about these formats but I also thought FBX was proprietary?
So it's still possible to write those files all by using Director?

Do you think it can be imported in Blender?
There is an import for FBX but not sure what version.
I would be very interested in FBX support but I only use Blender and I don't know which version they use. (you can also export to FBX but I'm not able to find what FBX version it is).
If you write it, can you maybe test it in Blender?

![afbeelding](https://user-images.githubusercontent.com/122493908/212206093-19b1383a-cad2-4b2b-9885-1d6212c98379.png)

**Comment by tomysshadow** [comment]
Well basically, Director allows you to create Xtras, which are third party extensions to Director written in a native programming language (such as C/C++.) So theoretically I could add the FBX SDK to Lingo that way. That would be the ideal for Shockwave 3D Worlds because they are normally exported from 3DS Max which also uses FBX, so the two formats are quite similar.

Yes, FBX is still proprietary, but it's widely supported by 3D modeling applications, so considering how similar it is to the Shockwave 3D World format, it would likely be the best choice for export format. But that definitely won't happen for a long time yet, because this project isn't high priority for me compared to other things I'm working on.

**Comment by tomysshadow** [comment]
This went a bit off topic, but the problem is resolved so I'm closing this issue.
